ctlearn_optimizer/common.py

```diff
@@ -0,0 +1,382 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import numpy as np
+import sklearn.metrics
+import os
+import re
+import yaml
+from ctlearn.run_model import run_model
+from multiprocessing import Pool
+
+
+def set_value(dictionary, value, *keys):
+    """Auxiliar function that modifies the value the keys point to in a nested dictionary
+
+    Dictionary can be a nested dictionary containing lists, this lists can also contain
+    nested dictionaries, and so on. *keys list can contain strings (which refer
+    dictionary keys) and integers (which refer list indices). Dictionary cannot be empty
+
+    Args:
+        dictionary: dictionary that contais the key-value pair
+        value: value to set
+        *keys: list of keys containing strings and integers
+
+    Returns:
+        modified dictionary
+
+    Example:
+        dictionary = {'a':[0,{'b':1},0]}
+        value = 2
+        *keys = ['a', 1, 'b']
+        set_value(dictionary, value, *keys) = {'a':[0,{'b':2},0]}
+    """
+    if type(dictionary) is not dict:
+        raise TypeError('set_value expects dict as first argument.')
+
+    _keys = keys[:-1]
+    _element = dictionary
+    for key in _keys:
+        _element = _element[key]
+    _element[keys[-1]] = value
+
+    return dictionary
+
+def create_nested_item(dictionary, *keys):
+    """Auxiliar function that creates an empty item with specific keys and positions in a nested dictionary
+
+    Dictionary can be a nested dictionary containing lists, this lists can also contain
+    nested dictionaries, and so on. *keys list can contain strings (which refer
+    dictionary keys) and integers (which refer list indices). Dictionary may or
+    may be not empty.
+
+    Args:
+        dictionary: dictionary to modify
+        *keys: list of keys containing strings and integers
+
+    Returns:
+        modified dictionary
+
+    Example:
+        dictionary = {}
+        value = 2
+        *keys = ['a', 'b', 1, 'c', 2 , 'd']
+        create_nested_item(dictionary, *keys) = {'a': {'b': [0, {'c': [0, 0, {'d': {}}]}]}}
+    """
+
+    if type(dictionary) is not dict:
+        raise TypeError('create_nested_item() expects dict as first argument.')
+
+    _keys = keys
+    _element = dictionary
+
+    #iterate over the list of keys
+    for n in range(len(_keys)):
```

Comment:
these for loops can likely be vectorized.
Also, the current structure of the function is a bit difficult to follow